servo/connectors/kubernetes.py

```diff
@@ -1487,7 +1487,7 @@ async def get_latest_pods(self) -> List[Pod]:
                 for ownRef in rs.metadata.owner_references)]
         if not rs_list:
             raise servo.ConnectorError('Unable to locate replicaset(s) for deployment "{self.name}"')
-        latest_rs = sorted(rs_list, key= lambda rs: rs.metadata.resource_version, reverse=True)[0]
+        latest_rs = sorted(rs_list, key= lambda rs: int(rs.metadata.annotations['deployment.kubernetes.io/revision']), reverse=True)[0]
```

Comment:
Can we use `observed_generation` instead of going to the annotations like this? (Refs: https://stackoverflow.com/questions/47100389/what-is-the-difference-between-a-resourceversion-and-a-generation, https://thenewstack.io/kubernetes-deployments-work/)

They should be equal to the same value other than being in a string on the annotation vs. an `int` but it feels more idiomatic/maintainable to utilize an API model affordance versus mining shit out of the annotations

Reply:
We should be able to handle the case when this annotation is missing - I don't think there is a guarantee it will be there. In any case, at least some protection should be avalable (even if it doesn't work, it can fail gracefully and indicate the error)

Reply:
Unfortunately that will not work in the context of replicasets. observedGeneration is equal to the deployment revision **on the Deployment object** but this does not hold true for the ReplicaSet object.

From https://thenewstack.io/kubernetes-deployments-work/

> As you can see, the ReplicaSet has indeed the label run=my-nginx. Not only that, the metadata contains again the deployment revision number. This is how a Deployment knows which ReplicaSet is the current one.

Further you can look at the yaml from this issue comment and see that the deployment.kubernetes.io/revision does not match the generation or observedGeneration of the replicasets

https://github.com/kubernetes/kubernetes/issues/41641#issuecomment-281064331

Reply:
Ack. I'm not fully up to speed on `ReplicaSet` and `Deployment` nuance yet.

So the `ReplicaSetStatus` object doesn't align with the annotation?

Reply:
No it does not

Reply:
Okay, I'm cool to merge then.

Are there cases where the annotation might not exist? Like should we defensively use `rs.metadata.annotations.get('deployment.kubernetes.io/revision', 0)`?

Reply:
That annotation should be guaranteed to exist on any replicaset managed by a deployment. Returning 0 as a default could lead to some inconsistent and difficult to debug behavior so I am opting to fail gracefully per Peter's request:

```
        if missing_revision_rsets := filter(lambda rs: 'deployment.kubernetes.io/revision' not in rs.metadata.annotations, rs_list):
            raise servo.ConnectorError(
                f'Unable to determine latest replicaset for deployment "{self.name}" due to missing revision annotation in replicaset(s)'
                f' "{", ".join(list(map(lambda rs: rs.metadata.name, missing_revision_rsets)))}"'
            )
```

Reply:
Boom. I'm onboard with this approach